dev/src/bulk-writer.ts

```diff
@@ -211,25 +224,92 @@ class BulkCommitBatch {
     // Capture the error stack to preserve stack tracing across async calls.
     const stack = Error().stack!;
 
-    return this.writeBatch.bulkCommit().catch(err => {
-      throw wrapError(err, stack);
-    });
+    let originalIndexMap: Map<number, number> = new Map(
+      Array.from(new Array(this.opCount), (_, i) => [i, i])
+    );
+
+    let retryAttempts = 0;
+    while (retryAttempts < MAX_RETRY_ATTEMPTS) {
+      let retryIndexes: number[] = [];
+      await this.backoff.backoffAndWait();
+      try {
+        const results = await this.writeBatch.bulkCommit();
+        retryIndexes = this.processResults(originalIndexMap, results);
+      } catch (err) {
```

Comment:
I tried to see if I could get rid of this re-assignment and got completely carried away. Here is a cleanup:

https://gist.github.com/schmidt-sebastian/702ba33ce0c3515d096089226a762934

Use at your own risk and take at your own consideration :)

Reply:
Huge fan of this organization. I implemented what you did, but changed `keys` to `docPaths` since I was a bit confused about what the keys were at first. I think this approach will work with the UpdateBuilder pattern on java as well.